File: backend/utils/throttle.js
import dotenv from "dotenv";
dotenv.config({ path: "./.env" });

let lastRequestTime = 0;
let isLocked = false;
const REQUEST_DELAY = parseInt(process.env.REQUEST_DELAY_MS) || 5000;

export async function throttleRequest() {
  // Wait if another request is currently throttling
  while (isLocked) {
    await new Promise(resolve => setTimeout(resolve, 100));
  }
  
  isLocked = true;
  try {
    const now = Date.now();
    const timeSinceLastRequest = now - lastRequestTime;
    
    if (timeSinceLastRequest < REQUEST_DELAY) {
      const waitTime = REQUEST_DELAY - timeSinceLastRequest;
      console.log(`⏳ [BETA-THROTTLE-V2] Waiting ${waitTime}ms for API limits...`);
      await new Promise(resolve => setTimeout(resolve, waitTime));
    }
    
    lastRequestTime = Date.now();
  } finally {
    isLocked = false;
  }
}

/**
 * Resets the throttle timer, useful after a 429 error to force a longer wait.
 * @param {number} additionalDelay - Extra time to wait in ms.
 */
export function forceBackoff(additionalDelay = 5000) {
    lastRequestTime = Date.now() + additionalDelay;
}

File: backend_fastapi/app/utils/throttle.py
import asyncio
import time
from app.config import settings

_last_request_time = 0.0
_lock = asyncio.Lock()

def get_request_delay() -> float:
    return settings.REQUEST_DELAY_MS / 1000.0

async def throttle_request():
    global _last_request_time
    async with _lock:
        request_delay = get_request_delay()
        now = time.time()
        time_since_last = now - _last_request_time
        if time_since_last < request_delay:
            wait_time = request_delay - time_since_last
            print(f"⏳ [BETA-THROTTLE-V2] Waiting {int(wait_time * 1000)}ms for API limits...")
            await asyncio.sleep(wait_time)
        _last_request_time = time.time()

def force_backoff(additional_delay: float = 5.0):
    global _last_request_time
    _last_request_time = time.time() + additional_delay
